src/engine/ebur128analysis.cpp

```diff
@@ -0,0 +1,381 @@
+  void operator()(GstSample *s) const { gst_sample_unref(s); };
+};
+
+struct FrameFormat {
+  enum class DataFormat {
+    S16,
+    S32,
+    FP32,
+    FP64
+  };
+
+  int channels;
+  int samplerate;
+  DataFormat format;
+
+  explicit FrameFormat(GstCaps *caps);
+};
+
+class EBUR128AnalysisImpl {
+  EBUR128AnalysisImpl() = default;
+
+ public:
+  static std::optional<EBUR128Measures> Compute(const Song &song);
+
+ private:
+  GstElement *convert_element_;
+
+  std::optional<FrameFormat> dsc;
+  QBuffer buffer_;
+
+  std::unique_ptr<ebur128_state, ebur128_state_deleter> state;
+
+  void InitState();
+  void AddFrames(const QByteArray &data);
+  std::optional<EBUR128Measures> finalize() const;
+
+  static void NewPadCallback(GstElement *elt, GstPad *pad, gpointer data);
+  static GstFlowReturn NewBufferCallback(GstAppSink *app_sink, gpointer self);
+};
+
+FrameFormat::FrameFormat(GstCaps *caps) {
+  GstStructure *structure = gst_caps_get_structure(caps, 0);
+  QString format_str = gst_structure_get_string(structure, "format");
+  gst_structure_get_int(structure, "rate", &samplerate);
+  gst_structure_get_int(structure, "channels", &channels);
+
+  if (format_str == "S16LE") {
+    format = DataFormat::S16;
+  }
+  else if (format_str == "S32LE") {
+    format = DataFormat::S32;
+  }
+  else if (format_str == "F32LE") {
+    format = DataFormat::FP32;
+  }
+  else if (format_str == "F64LE") {
+    format = DataFormat::FP64;
+  }
+  else {
+    qLog(Error) << "EBUR128AnalysisImpl: got unexpected format " << format_str;
+    Q_ASSERT(false && "Unexpected format. How did you get here?");
+  }
+}
+
+bool operator==(const FrameFormat &lhs, const FrameFormat &rhs) {
+
+  return std::tie(lhs.channels, lhs.samplerate, lhs.format) == std::tie(rhs.channels, rhs.samplerate, rhs.format);
+
+}
+bool operator!=(const FrameFormat &lhs, const FrameFormat &rhs) {
+
+  return !(lhs == rhs);
+
+}
+
+void EBUR128AnalysisImpl::InitState() {
+
+  Q_ASSERT(dsc);
+
+  state.reset(ebur128_init(dsc->channels, dsc->samplerate, EBUR128_MODE_I | EBUR128_MODE_LRA));
+  Q_ASSERT(state);
+
+}
+
+void EBUR128AnalysisImpl::AddFrames(const QByteArray &data) {
+
+  Q_ASSERT(dsc);
+  Q_ASSERT(state);
+
+  int bytesPerSample;
+  switch (dsc->format) {
+    case FrameFormat::DataFormat::S16:
+      bytesPerSample = sizeof(int16_t);
+      break;
+    case FrameFormat::DataFormat::S32:
+      bytesPerSample = sizeof(int32_t);
+      break;
+    case FrameFormat::DataFormat::FP32:
+      bytesPerSample = sizeof(float);
+      break;
+    case FrameFormat::DataFormat::FP64:
+      bytesPerSample = sizeof(double);
+      break;
+  }
+
+  int bytesPerFrame = dsc->channels * bytesPerSample;
+  Q_ASSERT(data.size() % bytesPerFrame == 0);
+  auto numFrames = data.size() / bytesPerFrame;
+
+  int ebur_error;
+  switch (dsc->format) {
+    case FrameFormat::DataFormat::S16:
+      ebur_error = ebur128_add_frames_short(&*state, reinterpret_cast<const int16_t *>(data.data()), numFrames);
```

Comment:
```suggestion
      ebur_error = ebur128_add_frames_short(&*state, reinterpret_cast<const int16_t*>(data.data()), numFrames);
```

Reply:
Same for the others below

Reply:
Hm, `.clang-format` says: `PointerAlignment: Right`.

Reply:
Point alignments should be right, except where there is no variable name, and like in casts. I see no option for this in clang-format.

Reply:
I see.
Non-automatic code formatting lives on. Oh well. :(